### ISSUE TITLE
Set gateway address + outbound NAT translation target via CLI

### ISSUE THREAD

**Issue by vlcty** [opnsense/core#4159]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[X] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md

[X] I have searched the existing issues and I'm convinced that mine is new.

**Is your feature request related to a problem? Please describe.**
Im building an OpenVPN tunnel to a public VPN provider which uses dynamic net30 addresing for the inside tunnel network. I'm not pulling a default route. I want to selectively route clients on my network over the tunnel. I've done this by creating a new gateway (set to the negotiated gateway after the tunnel was built) and an outgoing NAT rule (set to the negotiated tunnel address) to get it to work. This however is manual work I've to redo every time the tunnel is newly established. I'd like to be able to set the gateway address and NAT translation target via CLI so I could execute commands via the OpenVPN "--up /path/to/script" mechanism. 

**Describe the solution you'd like**
I think using configd actions would be great

**Describe alternatives you've considered**
There were none

**Additional context**
I've [created a pull request](https://github.com/opnsense/core/pull/4130) implementing the wanted features

**Comment by AdSchellevis** [comment]
You likely don’t have to add a hook for this, if you assign the OpenVPN interface it should receive a gateway automatically if I’m not mistaken.

Although I’m not against automating things in general, it should follow the same logic rules as the enduser input, which is why legacy code is more difficult to extend (no uniform model which validates input).

You can always pack your extension into a custom plugin for your own installations by the way, I don’t think we should offer this functionality out of the box (the risk of unwanted side affects is too high and it’s very likely not needed to outbound nat some nets over your tunnel)

**Comment by vlcty** [comment]
Assigning an interface does not automatically create a gateway. Especially when you can't configure the interfaces with any IP information. The OpenVPN server at the provider will assign a non static /30 every time the tunnel is created. The VPN providers does not know which networks are behind the tunnel. That's why ou need to NAT the source address to your local tunnel address.

**Comment by AdSchellevis** [comment]
Maybe it’s better to explain your setup on the forum, I’m not able to check anything out at the moment.

Given that many people use outbound nat on tunnel networks, I expect someone can point you in the right direction if you explain what you want to achieve.

**Comment by vlcty** [comment]
My setup is fine and everything works as expected. I've dony my research and testing prior and you can't do it with board tools at the moment. I'll explain with an example:

OPNsense builds a tunnel to ExpressVPN. The OpenVPN protocol negotiates 10.50.99.0/30 to be used as transfer network.   
The VPN server assignes 10.50.99.1 on its interface.   
The OPNsense assignes 10.50.99.2 on its interface.

You need to create a gateway let's say GW_ExpressVPN and set it to 10.50.99.1. The outbound NAT rules translation target must be 10.50.99.2. You can't tell a public VPN provider which network you use. There would be many conflicts. Now you can selectively route devices over that tunnel by creating a firewall rule and set GW_ExpressVPN as gateway. However when the tunnel is re-established a new /30 network will be negotiated breaking your setup. Manual changes need to be done.

With my pull request you can automate the task to change the IP of an existing gateway and the translation target for an outbound NAT rule by using configd. The same thing you do over GUI can now be made from CLI. IMHO there is no expected conflict with regular/existing setups.

**Comment by AdSchellevis** [comment]
I don’t think we should add it indeed, but further I’m also pretty sure it’s not needed to hook in a new gateway like this. If the wrong gateway is provided, your issue sounds similar to https://github.com/opnsense/core/pull/4134

About validation, your solution would accept (faulty) other input and saves it into the configuration, which could lead to all sorts of unintended side affects.

If you’re not able to receive a dynamic gateway at the moment with OpenVPN client, there might be another issue.

In some cases you can send traffic to the tunnel device without knowing it’s counterpart ip, for wireguard this works for example (there’s a switch on the interface for it). Theoretically you should be able to use the interface ip in your nat rule, to avoid static addresses.

Unfortunately we don’t have a lot of documentation available on our docs for this topic. We do have documentation about dynamic gateways inline, which might help understand why updating from within OpenVPN is in our vision the wrong way around (see our docs).

Long story short, our forum is a better place to discuss community support questions, if you’re convinced your setup is perfect, maybe better wrap the code into a plugin for personal use.
